ord_schema/validations.py

```diff
@@ -0,0 +1,428 @@
+
+@return_message_if_valid
+def ValidateReactionProvenance(message):
+    # Prepare datetimes
+    if message.experiment_start.value:
+        experiment_start = dateparser.parse(message.experiment_start.value)
+    if message.record_created.value:
+        record_created = dateparser.parse(message.record_created.value)
+    if message.record_modified.value:
+        record_modified = dateparser.parse(message.record_created.value)
+    # Check if record_created undefined
+    if message.record_modified.value and not message.record_created.value:
+        raise ValidationWarning('record_created not defined, but ' \
+            'record_modified is')
+    # Check signs of time differences
+    if message.experiment_start.value and message.record_created.value:
+        if (record_created - experiment_start).total_seconds() < 0:
+            raise ValueError('Record creation time should be after experiment')
+    if message.record_modified.value and message.record_created.value:
+        if (record_modified - record_created).total_seconds() < 0:
+            raise ValueError('Record modified time should be after creation')
+    # TODO(ccoley) could check if publication_url is valid, etc.
+
+@return_message_if_valid
+def ValidatePerson(message):
+    # NOTE(ccoley): final character is checksum, but ignoring that for now
+    if message.orcid:
+        if not re.match('[0-9]{4}-[0-9]{4}-[0-9]{4}-[0-9]{3}[0-9X]', 
+                message.orcid):
+            raise ValueError('Invalid ORCID: Enter as 0000-0000-0000-0000')
+
+@return_message_if_valid
+def ValidateTime(message):
+    ensure_float_nonnegative(message, 'value')
+    ensure_float_nonnegative(message, 'precision')
+    ensure_units_specified_if_value_defined(message)
+
+@return_message_if_valid
+def ValidateMass(message):
+    ensure_float_nonnegative(message, 'value')
+    ensure_float_nonnegative(message, 'precision')
+    ensure_units_specified_if_value_defined(message)
+
+@return_message_if_valid
+def ValidateMoles(message):
+    ensure_float_nonnegative(message, 'value')
+    ensure_float_nonnegative(message, 'precision')
+    ensure_units_specified_if_value_defined(message)
+
+@return_message_if_valid
+def ValidateVolume(message):
+    ensure_float_nonnegative(message, 'value')
+    ensure_float_nonnegative(message, 'precision')
+    ensure_units_specified_if_value_defined(message)
+
+@return_message_if_valid
+def ValidateConcentration(message):
+    ensure_float_nonnegative(message, 'value')
+    ensure_float_nonnegative(message, 'precision')
+    ensure_units_specified_if_value_defined(message)
+
+@return_message_if_valid
+def ValidatePressure(message):
+    ensure_float_nonnegative(message, 'value')
+    ensure_float_nonnegative(message, 'precision')
+    ensure_units_specified_if_value_defined(message)
+
+@return_message_if_valid
+def ValidateTemperature(message):
+    if message.units == message.CELSIUS:
+        ensure_float_range(message, 'value', -273.15, inf)
```

Comment:
Optional: you could have `ensure_float_range` include defaults for `min` and `max` so you don't need to specify both.